pkg/submitter/submitter.go

```diff
@@ -0,0 +1,504 @@
+		},
+			[]string{"slot"},
+		),
+		submitReward: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_reward",
+			Help:        "The current reward in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+	}
+
+	return submitter, submitter.resultCh
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) Stop() {
+	if s.lastSubmitCncl != nil {
+		s.lastSubmitCncl()
+	}
+	s.close()
+	level.Info(s.logger).Log("msg", "submitter shutdown complete")
+}
+
+func (s *Submitter) blockUntilTimeToSubmit(ctx context.Context) {
+	var (
+		lastSubmit time.Duration
+		timestamp  *time.Time
+		err        error
+	)
+	for {
+		select {
+		case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
+		default:
+		}
+		lastSubmit, timestamp, err = s.lastSubmit()
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
+	var (
+		timeToSubmit context.Context
+		cncl         context.CancelFunc
+	)
+	if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+		level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
+		timeToSubmit, cncl = context.WithDeadline(ctx, timestamp.Add(15*time.Minute))
+		defer cncl()
+	}
+	select {
+	case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
+	case <-timeToSubmit.Done(): // 15min since last submit has passed to can unblock.
+	}
+}
+
+func (s *Submitter) handleSubmit(ctx context.Context, result *mining.Result) {
+	go func(ctx context.Context, result *mining.Result) {
+		ticker := time.NewTicker(15 * time.Second)
+		for {
+			select {
+			case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
+				return
+			default:
+				profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
+				if err != nil {
+					level.Error(s.logger).Log("msg", "submit solution profit check", "err", err)
+					<-ticker.C
+					continue
+				}
+				if s.cfg.Mine.ProfitThreshold > 0 {
+					if profitPercent < int64(s.cfg.Mine.ProfitThreshold) {
+						level.Debug(s.logger).Log("msg", "transaction not profitable, so will wait for the next cycle")
+						<-ticker.C
+						continue
+					} else { // Transaction is profitable.
+						if statusID := s.getMinerStatus(); statusID != 1 { // I think status ID 3 was the status that allows to submit, but not sure need to double check the contract.
```

Comment:
That comment was for you to check the tellor3 contracts and check the actual status number :)

I think here maybe add a function for the log.
getStatusName(statusID) so that instead of a name we print the status string - not staked, in dispute etc.

```suggestion
						if statusID := s.getMinerStatus(); statusID != 1 { // I think status ID 3 was the status that allows to submit, but not sure need to double check the contract.
```

Reply:
Oh, sorry! I didn't realized that. I will check the contract then.